File: greedy/advantage-shuffle/advantage-shuffle.go
package advantageshuffle

import "sort"

func advantageCount(A []int, B []int) []int {
	sort.Ints(A)
	n := len(A)

	pq := make([]int, n)
	for i := range B {
		pq[i] = i
	}

	sort.Slice(pq, func(i, j int) bool {
		return B[pq[i]] > B[pq[j]]
	})

	res := make([]int, n)

	low, high := 0, n-1

	for i := range pq {
		index, value := pq[i], B[pq[i]]

		if A[high] > value {
			res[index] = A[high]
			high--
		} else {
			res[index] = A[low]
			low++
		}
	}
	return res
}

File: greedy/advantage-shuffle/advantage-shuffle.ts
function advantageCount(A: number[], B: number[]): number[] {
  A.sort((a, b) => a - b);
  const n: number = A.length;
  const res: number[] = new Array(n);

  const pq: number[] = [];

  for (let i = 0; i < n; i++) {
    pq.push(i);
  }

  pq.sort((a, b) => B[b] - B[a]);

  let low = 0;
  let high = n - 1;

  for (let i = 0; i < n; i++) {
    const index: number = pq[i];
    const value: number = B[index];

    if (A[high] > value) {
      res[index] = A[high];
      high--;
    } else {
      res[index] = A[low];
      low++;
    }
  }
  return res;
}
